src/components/contract-boost-calculator.tsx

```diff
@@ -1,484 +1,632 @@
+	diliT4: '0',
+	boost: 'boost8',
+	fetchState: FetchState.IDLE,
+	fetchRetryIn: 0,
+	doubleDuration: false,
+	baseIhr: '7440',
+	hatcheryCalm: '20',
+});
+
+const Calculator = generateCalculator<CalcData>(defaultCalcData());
+const { Input, Output } = Calculator;
+
+type FetchCoopDataProps = { readonly children: React.ReactNode };
+const FetchCoopDataButton = ({ children }: FetchCoopDataProps) => {
+	const apiUri = useContext(ApiUriContext);
+	const { data, updateData } = useContext<WithSetter<CalcData>>(Calculator.Context);
+
+	const fetchData = useCallback(async () => {
+		const backoffs = [1, 2, 5, 8, 13, 0] as const;
+		let eIResponse: EIBackupResponse | null = null;
+		for (const backoff of backoffs) {
+			updateData({ fetchState: FetchState.PENDING });
+			const rawEIResponse = await fetch(`${apiUri}/backup?EID=${data.eid}`);
+
+			if (rawEIResponse.ok) {
+				eIResponse = await rawEIResponse.json();
+				break;
+			} else if (backoff > 0) {
+				updateData({ fetchState: FetchState.RETRY, fetchRetryIn: backoff });
+				// we're doing this as a bunch of 1-second sleeps to get a countdown text
+				// this is technically not that accurate and it should be a setInterval
+				// but that was gnarly to write out
+				for (let remaining = backoff; remaining > 0; remaining--) {
+					updateData({ fetchRetryIn: remaining });
+					// no-loop-func is worried about `setTimeout` here ಠ_ಠ
+					// eslint-disable-next-line @typescript-eslint/no-loop-func
+					await new Promise((resolve) => void setTimeout(resolve, 1_000));
+				}
+			}
+		}
+
+		if (!eIResponse) {
+			updateData({ fetchState: FetchState.FAILURE });
+			return;
+		}
+
+		const resolveItemId = (itemId: number): SlottedArtifact => {
+			const items = eIResponse.artifactsDb.inventoryItemsList;
+			let lo = 0;
+			let hi = items.length;
+			while (lo <= hi) {
+				const mid = Math.floor((lo + hi) / 2);
+				const item = items[mid]!;
+				if (item.itemId === itemId) {
+					return item.artifact;
+				} else if (item.itemId > itemId) {
+					hi = mid - 1;
+				} else {
+					lo = mid + 1;
+				}
+			}
 
-	const getEquippedArtifactById = useCallback(
-		(artifactId: number) => {
-			const artifactData = equippedArtifactsListForSelectedIGN?.equippedArtifactsList?.find(
-				(artifact: EquippedArtifact) => artifact.spec.name === artifactId,
+			throw new Error(
+				`Egg, Inc. backup contained an item ID (${itemId}) that is missing from itself`,
 			);
+		};
+
+		const ihrForSet = (set: SlottedArtifact[]) => {
+			const life = [0, 0, 0];
+			let chalice = 1;
+			for (const artifact of set) {
+				for (const stone of artifact.stonesList) {
+					if (stone.name === ArtifactSpec.Name.LIFE_STONE) {
+						life[stone.level]! += 1;
+					}
+				}
+
+				if (artifact.spec.name === ArtifactSpec.Name.THE_CHALICE) {
+					chalice = chaliceMultiplier(artifact.spec);
+				}
+			}
+
+			return {
+				stones: life,
+				effect: chalice * 1.02 ** life[0]! * 1.03 ** life[1]! * 1.04 ** life[2]!,
+			};
```

Comment:
i feel like this could be a little more descriptive by reading the array as let [x, y, z] = [0, 0, 0] rather than accessing the index of the life array

Reply:
same with dili

Reply:
i dig it, coming soon to a commit near you!